overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	pattern = duplicateSlashes.ReplaceAllString(pattern, `${1}/`)
+	pattern = charsDoublestar.ReplaceAllString(pattern, `${1}*`)
+	pattern = doublestarChars.ReplaceAllString(pattern, `${1}*${2}`)
+	pattern = duplicateDoublestar.ReplaceAllString(pattern, `/**`)
+	pattern = starsAnyMaybeStars.ReplaceAllStringFunc(pattern, shiftStarAfterAnyChars)
+	if strings.HasSuffix(pattern, "/**/*") {
+		// Strip trailing "/*" from suffix
+		return pattern[:len(pattern)-2]
+	}
+	return pattern
+}
+
+func shiftStarAfterAnyChars(s string) string {
+	return strings.Map(deleteStars, s) + "*"
+}
+
+func deleteStars(r rune) rune {
+	if r == '*' {
+		return -1
+	}
+	return r
+}
+
+type priorityType int
+
+const (
+	worstPriority priorityType = iota
+	priorityGlobDoublestar
+	priorityTerminalDoublestar
+	priorityDoublestar
+	priorityGlob
+	prioritySinglestar
+	prioritySingleChar
+	priorityTerminated
+	priorityLiteral
+)
+
+type nextPatternsContainer struct {
+	currPriority    priorityType
+	nextPatternsMap map[string]int
+}
+
+func (np *nextPatternsContainer) addWithPriority(priority priorityType, pattern string, e int) {
+	if priority < np.currPriority {
+		return
+	}
+	if priority > np.currPriority {
+		np.nextPatternsMap = make(map[string]int)
+		np.currPriority = priority
+	}
+	np.nextPatternsMap[pattern] = e
+}
+
+func (np *nextPatternsContainer) nextPatterns() map[string]int {
+	return np.nextPatternsMap
+}
+
+// Determines which of the given path patterns is the most specific (top priority).
+//
+// Assumes that all of the given patterns satisfy ValidatePathPattern(), so this
+// is not verified as part of this function. Additionally, also assumes that the
+// patterns have been previously expanded using ExpandPathPattern(), so there
+// are no groups in any of the patterns.
+//
+// Below are some sample patterns, in order of precedence, though precedence is
+// only guaranteed between two patterns which may match the same path:
+//
+//	# literals
+//	- /foo/bar/baz
+//	- /foo/bar/
+//	# terminated
+//	- /foo/bar
+//	# any single character
+//	- /foo/bar?baz
+//	- /foo/bar?
+//	- /foo/bar?/
+//	# singlestars
+//	- /foo/bar/*/baz
+//	- /foo/bar/*/
+//	- /foo/bar/*/*baz
+//	- /foo/bar/*/*
+//	- /foo/bar/*
+//	- /foo/bar/*/**
+//	# glob
+//	- /foo/bar*baz
+//	- /foo/bar*/baz
+//	- /foo/bar*/baz/**
+//	- /foo/bar*/
+//	- /foo/bar*/*baz
+//	- /foo/bar*/*/baz
+//	- /foo/bar*/*/
+//	- /foo/bar*/*
+//	- /foo/bar*
+//	# doublestars
+//	- /foo/bar/**/baz
+//	- /foo/bar/**/*baz/
+//	- /foo/bar/**/*baz
+//	# terminal doublestar
+//	- /foo/bar/**/        # These are tough... usually, /foo/bar/**/ would have precedence over
+//	- /foo/bar/**/*       # precedence over /foo/bar/**/*baz, but in this case,
+//	- /foo/bar/**         # the trailing *baz adds more specificity.
+//	# glob with immediate doublestar
+//	- /foo/bar*/**/baz
+//	- /foo/bar*/**/
+//	- /foo/bar*/**
+func GetHighestPrecedencePattern(patterns []string) (string, error) {
```

Comment:
Is it necessary to both sort and obtain the max priority pattern? If you only need one, would maintaining some `maxPriorityPattern` each iteration work?

Reply:
There could be multiple patterns with the same priority at a given iteration, so all top priority patterns need to be stored until the next iteration, when we look at the next character. I don't think sorting comes into play here.

Technically, the patterns could probably be sorted to get a good idea of which probably has the highest priority, but with escaped characters and the special cases around singlestars and doublestars, this would not really help and might distract from the checks which really matter.

I'm not fully sure I understand what you mean though